R/extract.R

```diff
@@ -0,0 +1,144 @@
+#'  importance/explainers.
+#'
+#' However, users should not invoke the `predict()` method on an extracted
+#'  model. There may be preprocessing operations that `workflows` has executed on
+#'  the data prior to giving it to the model. Bypassing these can lead to errors
+#'  or silently generating incorrect predictions.
+#'
+#' *Good*:
+#' ```r
+#'    workflow_fit %>% predict(new_data)
+#' ```
+#'
+#' *Bad*:
+#' ```r
+#'    workflow_fit %>% extract_fit_engine()  %>% predict(new_data)
+#'    # or
+#'    workflow_fit %>% extract_fit_parsnip() %>% predict(new_data)
+#' ```
+#' @return
+#' The extracted value from the object, `x`, as described in the
+#' description section.
+#' @examples
+#' library(tune)
+#'
+#' extract_workflow_set_result(two_class_res, "none_cart")
+#'
+#' extract_workflow(two_class_res, "none_cart")
+#' @export
+extract_workflow_set_result <- function(x, id, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   y$result[[1]]
+}
+
+#' @export
+#' @rdname extract_workflow_set_result
+extract_workflow.workflow_set <- function(x, id, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   y$info[[1]]$workflow[[1]]
+}
+
+#' @export
+#' @rdname extract_workflow_set_result
+extract_spec_parsnip.workflow_set <- function(x, id, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   extract_spec_parsnip(y$info[[1]]$workflow[[1]])
+}
+
+
+#' @export
+#' @rdname extract_workflow_set_result
+extract_recipe.workflow_set <- function(x, id, estimated = TRUE, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   extract_recipe(y$info[[1]]$workflow[[1]], estimated = estimated)
+}
+
+#' @export
+#' @rdname extract_workflow_set_result
+extract_fit_parsnip.workflow_set <- function(x, id, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   extract_fit_parsnip(y$info[[1]]$workflow[[1]])
+}
+
+#' @export
+#' @rdname extract_workflow_set_result
+extract_fit_engine.workflow_set <- function(x, id, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   extract_fit_engine(y$info[[1]]$workflow[[1]])
+}
+
+#' @export
+#' @rdname extract_workflow_set_result
+extract_mold.workflow_set <- function(x, id, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   extract_mold(y$info[[1]]$workflow[[1]])
+}
+
+#' @export
+#' @rdname extract_workflow_set_result
+extract_preprocessor.workflow_set <- function(x, id, ...) {
+   check_id(x, id)
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   extract_preprocessor(y$info[[1]]$workflow[[1]])
+}
+
+# ------------------------------------------------------------------------------
+
+
+check_id <- function(x, id) {
+   if (length(id) != 1) {
+      rlang::abort("'id' should have a single value.")
+   }
+   y <- x %>% dplyr::filter(wflow_id == id[1])
+   if (nrow(y) != 1) {
+      halt("No workflow ID found for '", id[1], "'")
+   }
+   invisible(NULL)
+}
```

Comment:
Possibly:

```r
filter_id <- function(x, id) {
  if (!rlang::is_string(id)) {
    halt("`id` must be a single string.")
  }
  
  out <- dplyr::filter(x, wflow_id == id)
  
  if (nrow(out) != 1L) {
    halt("`id` must correspond to a single row in `x`.")
  }
  
  out
}
```

What I like about this is:
- It also checks the type of `id`, and ensure it isn't an `NA`
- It is possible that a bad workflowset could end up with 2 rows with the same ID. The error message here is just a little more informative in case that happens. You could also add a `nrow(x) == 0` case if you like the "no workflow id found" error message